File: src/services/contract/contractTransforms.ts
import { ContractMatchCreateRequest, ContractMatchMode, ContractScheduleVoteRequest } from "../../contracts/generated/contractModels";

// Note for non-coders: generated contract types describe raw API data, while helpers here hold app-specific business rules.
const ensureModeTeamSize = (mode: ContractMatchMode, team: string[], teamLabel: string) => {
  const expected = mode === "1v1" ? 1 : 2;
  if (team.length !== expected) {
    throw new Error(`${teamLabel} måste ha ${expected} spelare i läget ${mode}.`);
  }
};

export const buildMatchCreateRequest = (
  payload: Omit<ContractMatchCreateRequest, "match_mode" | "team1" | "team2"> & {
    team1: string[] | string;
    team2: string[] | string;
    match_mode?: ContractMatchMode;
  },
): ContractMatchCreateRequest => {
  const team1 = Array.isArray(payload.team1) ? payload.team1 : [payload.team1];
  const team2 = Array.isArray(payload.team2) ? payload.team2 : [payload.team2];
  const matchMode = payload.match_mode ?? (team1.length === 1 && team2.length === 1 ? "1v1" : "2v2");

  ensureModeTeamSize(matchMode, team1, "Lag 1");
  ensureModeTeamSize(matchMode, team2, "Lag 2");

  return {
    ...payload,
    team1,
    team2,
    match_mode: matchMode,
  };
};

export const buildScheduleVoteRequest = (payload: ContractScheduleVoteRequest): ContractScheduleVoteRequest => {
  if (payload.slot_preferences.length === 0) {
    throw new Error("Välj minst en tidslucka.");
  }
  return payload;
};

File: ios-native/PadelNative/Services/Contract/ContractTransforms.swift
import Foundation

// Note for non-coders:
// Generated contract models define the exact API shape, and this file keeps app-specific rules separate.
enum ContractTransforms {
    static func toContractMatchRequest(from submission: MatchSubmission) throws -> ContractMatchCreateRequest {
        let mode = submission.matchMode ?? inferredMode(team1Count: submission.teamAName.count, team2Count: submission.teamBName.count)
        try validate(teamCount: submission.teamAName.count, mode: mode, label: "Team A")
        try validate(teamCount: submission.teamBName.count, mode: mode, label: "Team B")

        return ContractMatchCreateRequest(
            team1: submission.teamAName,
            team2: submission.teamBName,
            team1IDs: submission.teamAPlayerIds.map { id in id.flatMap(UUID.init(uuidString:)) },
            team2IDs: submission.teamBPlayerIds.map { id in id.flatMap(UUID.init(uuidString:)) },
            team1Sets: submission.teamAScore,
            team2Sets: submission.teamBScore,
            createdBy: submission.createdBy,
            sourceTournamentId: submission.sourceTournamentId,
            matchMode: mode
        )
    }

    static func toContractVote(dayId: UUID, profileId: UUID, slotPreferences: [AvailabilitySlot]) throws -> ContractScheduleVoteRequest {
        if slotPreferences.isEmpty {
            throw APIError.requestFailed(statusCode: 400)
        }
        return ContractScheduleVoteRequest(
            pollDayId: dayId,
            profileId: profileId,
            slotPreferences: slotPreferences.map(\.contractValue)
        )
    }

    private static func inferredMode(team1Count: Int, team2Count: Int) -> ContractMatchMode {
        (team1Count == 1 && team2Count == 1) ? .oneVsOne : .twoVsTwo
    }

    private static func validate(teamCount: Int, mode: ContractMatchMode, label: String) throws {
        let expectedCount = mode == .oneVsOne ? 1 : 2
        if teamCount != expectedCount {
            throw NSError(domain: "ContractTransforms", code: 1, userInfo: [NSLocalizedDescriptionKey: "\(label) must have \(expectedCount) players for mode \(mode.rawValue)."])
        }
    }
}

private extension AvailabilitySlot {
    var contractValue: ContractAvailabilitySlot {
        switch self {
        case .morning: return .morning
        case .day: return .day
        case .evening: return .evening
        }
    }
}
